# **Clasificación**

### En este archivo trataremos con diferentes modelos de clasificación para predecir el ganador del torneo de fútbol de esta temporada 23-24.

Comenzamos importando todas las dependecias necesarias para el correcto funcionamiento del documento.

In [95]:
# Importamos librerías

# Visualización de datos
%matplotlib inline

# Manipulación de datos
import pandas as pd
import numpy as np

# Preparación de los datos
from sklearn.preprocessing import scale, LabelEncoder
from sklearn.model_selection import train_test_split

# Modelos de predicción
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier

# Métricas de evaluación
from sklearn.metrics import accuracy_score, classification_report

El siguiente paso es importar los datos con lo que vamos a trabajar.

In [96]:
partidos = pd.read_csv('../../data/partidos_limpio.csv')
partidos.head()

Season        Round  Day        Date Results           Home  \
0  2023-2024  Round of 16  Tue  2024-02-13       A     RB Leipzig   
1  2023-2024  Round of 16  Tue  2024-02-13       A  FC Copenhagen   
2  2023-2024  Round of 16  Wed  2024-02-14       H      Paris S-G   
3  2023-2024  Round of 16  Wed  2024-02-14       H          Lazio   
4  2023-2024  Round of 16  Tue  2024-02-20       D  PSV Eindhoven   

    Country (Home)  Points (Home)  Score (Home)  Score (Away)  ...  MP_away  \
0          Germany      88.736698             0             1  ...     10.0   
1          Denmark      80.431647             1             3  ...     10.0   
2           France     114.334580             2             0  ...      8.0   
3            Italy      99.943311             1             0  ...     10.0   
4  The Netherlands      98.784903             1             1  ...     10.0   

  Starts_away Gls_away Ast_away G+A_away  G-PK_away  PK_away  PKatt_away  \
0       110.0     20.0     17.0     37.0       20.0      0.0         1.0   
1       110.0     28.0     20.0     48.0       25.0      3.0         3.0   
2        88.0      8.0      5.0     13.0        8.0      0.0         1.0   
3       110.0     18.0     14.0     32.0       16.0      2.0         2.0   
4       110.0     15.0     12.0     27.0       14.0      1.0         1.0   

   CrdY_away  CrdR_away  
0       18.0        0.0  
1       10.0        0.0  
2       18.0        0.0  
3       13.0        1.0  
4       16.0        0.0  

[5 rows x 39 columns]

### Preparación de los datos.

Queremos columnas numéricas para nuestra predicción, así que pasaremos todas aquellas columnas categóricas a numéricas usando Label Encoder. Además guardamos estas transformaciones en un diccionario que llamamos 'mapping' para poder posteriormente deshacer los cambios con facilidad. 

La única columna no numérica que no cambiaremos será la columna 'Results' ya que esa será nuestra variable objetivo (y) y podemos dejarla como categórica.

In [97]:
# Columnas a modificar
cols = ['Season', 'Round', 'Day', 'Home', 'Away', 'Country (Home)', 'Country (Away)', 'Venue', 'Referee']

# Inicializamos el label encoder
label_encoder = LabelEncoder()

# Creamos un diccionario para guardar los mapeos
mapping = {}

# Iteramos sobre las columnas y las transformamos
for col in cols:
    # Concatenamos los valores necesarios
    if col in ['Home', 'Away']:
        if 'Squad' not in mapping:
            name = 'Squad'
            squad = pd.concat([partidos['Home'], partidos['Away']])
            label_encoder.fit(squad)      
    elif col in ['Country (Home)', 'Country (Away)']:
        if 'Country' not in mapping:
            name = 'Country'
            country = pd.concat([partidos['Country (Home)'], partidos['Country (Away)']])
            label_encoder.fit(country)
    else:
        name = col
        label_encoder.fit(partidos[col])
    
    # Transformamos los valores 
    partidos[col] = label_encoder.transform(partidos[col])
    
    # Creamos un mapeo de los valores
    mapping[name] = dict(zip(label_encoder.classes_, label_encoder.transform(label_encoder.classes_)))

# Transformamos la columna 'Date' a datetime
partidos['Date'] = pd.to_datetime(partidos['Date'])

# Separar la fecha en año, mes y día
partidos['Year'] = partidos['Date'].dt.year
partidos['Month'] = partidos['Date'].dt.month
partidos['Number Day'] = partidos['Date'].dt.day # Lo llamamos 'Number Day' para evitar confusiones con la columna 'Day' que ya existe

# Eliminamos la columna 'Date'
partidos.drop('Date', axis=1, inplace=True)

# Verificamos los cambios
partidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 598 entries, 0 to 597
Data columns (total 41 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Season          598 non-null    int32  
 1   Round           598 non-null    int32  
 2   Day             598 non-null    int32  
 3   Results         598 non-null    object 
 4   Home            598 non-null    int32  
 5   Country (Home)  598 non-null    int32  
 6   Points (Home)   598 non-null    float64
 7   Score (Home)    598 non-null    int64  
 8   Score (Away)    598 non-null    int64  
 9   Points (Away)   598 non-null    float64
 10  Country (Away)  598 non-null    int32  
 11  Away            598 non-null    int32  
 12  Venue           598 non-null    int32  
 13  Referee         598 non-null    int32  
 14  # Pl_home       538 non-null    float64
 15  Age_home        538 non-null    float64
 16  MP_home         538 non-null    float64
 17  Starts_home     538 non-null    flo

Eliminamos las filas con datos nulos.

In [98]:
partidos = partidos.dropna()

Ahora ya definimos nuestras variables para las predicciones. En nuestro caso, queremos predecir los resultados del partido, por lo que querremos predecir la columna 'Results'.

Eliminamos además las columnas 'Score (Home)' y 'Score (Away)' ya que en las predicciones nunca tendremos los resultados del partido, por lo que no tiene sentido entrenar los modelos con estas dos columnas. También eliminamos la columna 'Referee' ya que a la hora de predecir un partido tampoco sabremos quién será el árbitro del mismo.

In [99]:
X = partidos.drop(labels=['Results', 'Score (Home)', 'Score (Away)', 'Referee'], axis=1)
y = partidos['Results']

# Estandarizamos los datos

# Columnas a estandarizar
cols = [['Points (Home)', 'Points (Away)', '# Pl_home','Age_home','MP_home','Starts_home','Gls_home','Ast_home','G+A_home','G-PK_home','PK_home','PKatt_home','CrdY_home','CrdR_home','# Pl_away','Age_away','MP_away','Starts_away','Gls_away','Ast_away','G+A_away','G-PK_away','PK_away','PKatt_away','CrdY_away','CrdR_away']]

# Se recorren las columnas especificadas y se escala cada una
for col in cols:
    X[col] = scale(X[col])

Separamos en train y test las variables que usaremos para entrenar nuestros modelos.

In [100]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### Modelos de predicción.

Comenzamos con la Regresión Logística.

In [101]:
lr = LogisticRegression(random_state=42, max_iter=1000, C=1.0)
lr.fit(X_train, y_train)
lr_pred = lr.predict(X_test)

c:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Seguimos con SVC (Support Vector Classification).

In [102]:
svc = SVC(random_state = 912, kernel='rbf')
svc.fit(X_train, y_train)
svc_pred = svc.predict(X_test)

Nuestro último modelo de clasificación será XGBoost.

In [103]:
xgb = GradientBoostingClassifier(random_state=42)
xgb.fit(X_train, y_train)
xgb_pred = xgb.predict(X_test)

### Métricas de evaluación.

Para hacer estas métricas, haremos uso del informe de clasificación (Classification Report), una tabla que resume las métricas de evaluación del rendimiento de un modelo de clasificación en un conjunto de datos de prueba. 

¿Qué métricas aparecen en este informe?

- Precision (Precisión): La precisión indica la proporción de instancias clasificadas como positivas que son verdaderamente positivas. Se calcula como el número de verdaderos positivos dividido por el número total de predicciones positivas (verdaderos positivos más falsos positivos).

- Recall (Recuperación o Sensibilidad): El recall indica la proporción de instancias positivas que fueron correctamente clasificadas. Se calcula como el número de verdaderos positivos dividido por el número total de instancias positivas (verdaderos positivos más falsos negativos).

- F1-score (Puntaje F1): Es la media armónica de la precisión y el recall. Proporciona un equilibrio entre precisión y recall. Se calcula como 2 * (precision * recall) / (precision + recall).

- Support (Soporte): Es el número de muestras verdaderas que pertenecen a cada clase en el conjunto de datos de prueba.

- Accuracy (Exactitud): Es la proporción de muestras correctamente clasificadas entre todas las muestras. Se calcula como el número de muestras correctamente clasificadas dividido por el número total de muestras.

- Macro average (Promedio Macro): Es el promedio sin ponderar de las métricas para cada clase. Calcula las métricas independientemente para cada clase y luego toma el promedio sin tener en cuenta el desequilibrio en el tamaño de las clases.

- Weighted average (Promedio Ponderado): Es el promedio ponderado de las métricas para cada clase, donde cada clase contribuye con su peso relativo al número total de muestras en el conjunto de datos de prueba.

In [104]:
print("\nMétricas Regresión Logística:")
print("Precisión:", accuracy_score(y_test, lr_pred))
print("Informe de Clasificación:")
print(classification_report(y_test, lr_pred, zero_division=0))


Métricas Regresión Logística:
Precisión: 0.6111111111111112
Informe de Clasificación:
              precision    recall  f1-score   support

           A       0.60      0.70      0.65        46
           D       0.18      0.05      0.08        37
           H       0.66      0.82      0.73        79

    accuracy                           0.61       162
   macro avg       0.48      0.52      0.49       162
weighted avg       0.54      0.61      0.56       162



In [105]:
print("\nMétricas SVC:")
print("Precisión:", accuracy_score(y_test, svc_pred))
print("Informe de Clasificación:")
print(classification_report(y_test, svc_pred, zero_division=0))


Métricas SVC:
Precisión: 0.4876543209876543
Informe de Clasificación:
              precision    recall  f1-score   support

           A       0.00      0.00      0.00        46
           D       0.00      0.00      0.00        37
           H       0.49      1.00      0.66        79

    accuracy                           0.49       162
   macro avg       0.16      0.33      0.22       162
weighted avg       0.24      0.49      0.32       162



In [106]:
print("\nMétricas XGBoost:")
print("Precisión:", accuracy_score(y_test, xgb_pred))
print("Informe de Clasificación:")
print(classification_report(y_test, xgb_pred, zero_division=0))


Métricas XGBoost:
Precisión: 0.5617283950617284
Informe de Clasificación:
              precision    recall  f1-score   support

           A       0.57      0.52      0.55        46
           D       0.24      0.14      0.17        37
           H       0.63      0.78      0.70        79

    accuracy                           0.56       162
   macro avg       0.48      0.48      0.47       162
weighted avg       0.52      0.56      0.53       162



### Predicciones

Ahora hagamos las predicciones de los partidos que aún quedan por jugar.

In [107]:
# Definimos una lista llamada 'semis' que contiene datos de partidos de fútbol de las semifinales
semis = [['2023-2024', 'Semi-finals', 'Tue', 'Bayern Munich', 'Germany', 107.882298136646, 114.5545351473923, 'Spain', 'Real Madrid', 'Allianz Arena', 23, 28.3, 10, 110, 18, 14, 32, 16, 2, 2, 13.0, 1.0, 22, 28.0, 10, 110, 20, 17, 37, 20, 0, 1, 18.0, 0.0, 2024, 4, 30],
        ['2023-2024', 'Semi-finals', 'Wed', 'Dortmund', 'Germany', 91.17303312629399, 114.33458049886625, 'France', 'Paris S-G', 'Signal Iduna Park', 23, 28.0, 10, 110, 15, 12, 27, 14, 1, 1, 16.0, 0.0, 21, 25.3, 10, 110, 19, 12, 31, 16, 3, 3, 27.0, 0.0, 2024, 5, 1],
        ['2023-2024', 'Semi-finals', 'Tue', 'Paris S-G', 'France', 114.33458049886625, 91.17303312629399, 'Germany', 'Dortmund', 'Parc des Princes', 21, 25.3, 10, 110, 19, 12, 31, 16, 3, 3, 27.0, 0.0, 23, 28.0, 10, 110, 15, 12, 27, 14, 1, 1, 16.0, 0.0, 2024, 5, 7],
        ['2023-2024', 'Semi-finals', 'Wed', 'Real Madrid', 'Spain', 114.5545351473923, 107.882298136646, 'Germany', 'Bayern Munich', 'Estadio Santiago Bernabéu', 22, 28.0, 10, 110, 20, 17, 37, 20, 0, 1, 18.0, 0.0, 23, 28.3, 10, 110, 18, 14, 32, 16, 2, 2, 13.0, 1.0, 2024, 5, 8]]

# Obtenemos las columnas relevantes del DataFrame 'partidos' para usar como nombres de columnas en el DataFrame 'semis'
partidos_cols = partidos.drop(labels=['Results', 'Score (Home)', 'Score (Away)', 'Referee'], axis=1).columns

# Creamos un DataFrame 'semis' a partir de la lista 'semis' con las columnas obtenidas del DataFrame 'partidos'
semis = pd.DataFrame(semis, columns=(partidos_cols))

semis.head()

Season        Round  Day           Home Country (Home)  Points (Home)  \
0  2023-2024  Semi-finals  Tue  Bayern Munich        Germany     107.882298   
1  2023-2024  Semi-finals  Wed       Dortmund        Germany      91.173033   
2  2023-2024  Semi-finals  Tue      Paris S-G         France     114.334580   
3  2023-2024  Semi-finals  Wed    Real Madrid          Spain     114.554535   

   Points (Away) Country (Away)           Away                      Venue  \
0     114.554535          Spain    Real Madrid              Allianz Arena   
1     114.334580         France      Paris S-G          Signal Iduna Park   
2      91.173033        Germany       Dortmund           Parc des Princes   
3     107.882298        Germany  Bayern Munich  Estadio Santiago Bernabéu   

   ...  Ast_away  G+A_away  G-PK_away  PK_away  PKatt_away  CrdY_away  \
0  ...        17        37         20        0           1       18.0   
1  ...        12        31         16        3           3       27.0   
2  ...        12        27         14        1           1       16.0   
3  ...        14        32         16        2           2       13.0   

   CrdR_away  Year  Month  Number Day  
0        0.0  2024      4          30  
1        0.0  2024      5           1  
2        0.0  2024      5           7  
3        1.0  2024      5           8  

[4 rows x 37 columns]

Ahora, para poder aplicar nuestros modelos a estos nuevos datos, debemos convertirlos a numéricos. Para ello, aplicamos el mismo mapping que usamos anteriormente.

In [108]:
data = semis.copy()

# Aplicamos mapping a las columnas
for col, col_mapping in mapping.items():
    if col in data.columns:
        data[col] = data[col].map(col_mapping)
    else:
        if col == 'Squad':
            data['Home'] = data['Home'].map(col_mapping)
            data['Away'] = data['Away'].map(col_mapping)
        elif col == 'Country':
            data['Country (Home)'] = data['Country (Home)'].map(col_mapping)
            data['Country (Away)'] = data['Country (Away)'].map(col_mapping)

data.head()

Season  Round  Day  Home  Country (Home)  Points (Home)  Points (Away)  \
0      20      3    4     7               7     107.882298     114.554535   
1      20      3    5    16               7      91.173033     114.334580   
2      20      3    4    45               6     114.334580      91.173033   
3      20      3    5    50              13     114.554535     107.882298   

   Country (Away)  Away  Venue  ...  Ast_away  G+A_away  G-PK_away  PK_away  \
0              13    50      1  ...        17        37         20        0   
1               6    45     77  ...        12        31         16        3   
2               7    16     71  ...        12        27         14        1   
3               7     7     30  ...        14        32         16        2   

   PKatt_away  CrdY_away  CrdR_away  Year  Month  Number Day  
0           1       18.0        0.0  2024      4          30  
1           3       27.0        0.0  2024      5           1  
2           1       16.0        0.0  2024      5           7  
3           2       13.0        1.0  2024      5           8  

[4 rows x 37 columns]

Hacemos nuestras predicciones de las semifinales de este torneo

In [109]:
lr_pred_actual = lr.predict(data)
svc_pred_actual = svc.predict(data)
xgb_pred_actual = xgb.predict(data)

### Resultados

Veamos qué resultados obtenemos de cada modelo realizado.

Para visualizar los resultados haremos un DataFrame nuevo. Para crear este nuevo DataFrame haremos 4 listas: cada una corresponde a una columna del nuevo DataFrame.

In [110]:
# Se obtiene la lista de los valores de las columnas 'Squad' y 'Rk' de cada predicción
X_home = semis['Home'].tolist()
X_away = semis['Away'].tolist()
lr_pred_lst = lr_pred_actual.tolist()
svc_pred_lst = svc_pred_actual.tolist()
xgb_pred_lst = xgb_pred_actual.tolist()

# Se crea un DataFrame con los valores obtenidos
res = pd.DataFrame({'Home': X_home, 'Away': X_away, 'LR': lr_pred_lst, 'SVC': svc_pred_lst, 'XGB': xgb_pred_lst})

# Mostrar los resultados
res

Home           Away LR SVC XGB
0  Bayern Munich    Real Madrid  A   H   D
1       Dortmund      Paris S-G  A   H   A
2      Paris S-G       Dortmund  A   H   D
3    Real Madrid  Bayern Munich  A   H   D

No queda claro qué equipos pasarían a la siguiente ronda. Por lo tanto, estudiaremos las probabilidades que se han obtenido a la hora de hacer las predicciones. 

Es decir, el modelo hace una predicción porque, por ejemplo, la opción 'A' le ha salido con un 90% de probabilidad y las otras dos opciones con menor probabilidad. Basándose en estos datos, en la pantalla se muestra aquella opción con mayor probabilidad, en el caso que hemos visto, se monstraría 'A' como predicción. Nosotros analizaremos estas predicciones y el equipo que tenga mayor probabilidad de ganar pasa a la siguiente ronda.

In [111]:
# Predicción de probabilidades para Regresión Logística (lr)
lr_pred_prob = lr.predict_proba(data)


# Predicción de probabilidades para SVC (svc)

# Calcular las puntuaciones de decisión para SVC (svc)
svc_decision_scores = svc.decision_function(data)
# Convertir las puntuaciones en probabilidades usando la función sigmoide
svc_pred_prob = 1 / (1 + np.exp(-svc_decision_scores))


# Predicción de probabilidades para XGBoost (xgb)
xgb_pred_prob = xgb.predict_proba(data)


# Crear un diccionario con las probabilidades de cada modelo
pred_probs = {}

# Se obtienen las claves, que corresponden a los valores únicos de 'y' (variable objetivo)
pred_probs_keys = y.unique()

# Se recorren las claves y se obtienen las probabilidades de cada modelo
for key in pred_probs_keys:
    pred_probs[key] = []
    for i in range(len(lr_pred_prob)):
        # Se obtiene el índice de la clave en el array de claves
        x = np.where(pred_probs_keys == key)[0][0]
        # Se añaden las probabilidades de cada modelo a la lista de la clave correspondiente
        pred_probs[key].append([lr_pred_prob[i][x], svc_pred_prob[i][x], xgb_pred_prob[i][x]])

print(pred_probs)

{'A': [[0.9992342079861941, 0.7314224353411514, 0.3278833058430229], [0.9979236052920363, 0.7313657919889971, 0.5220480870374677], [0.7224568609987156, 0.7312988180391108, 0.22986099837857438], [0.7742763397897036, 0.7313173990324053, 0.22505442690372432]], 'H': [[0.000765693039203509, 0.4446658105986785, 0.6657942770815379], [0.0020763227961652286, 0.44466523310524053, 0.45110358238428216], [0.2775261714349433, 0.4446683567050252, 0.7586002361562705], [0.2256753941121798, 0.44466760629762464, 0.7631860256349595]], 'D': [[9.897460237890221e-08, 0.9022113163656188, 0.006322417075439371], [7.19117983744838e-08, 0.9022143874367771, 0.026848330578250243], [1.696756634109644e-05, 0.9022166459084169, 0.011538765465155301], [4.826609811678876e-05, 0.902215978537091, 0.011759547461316027]]}


Una vez obtenidas las probabilidades, agregamos las más relevantes a nuestra tabla de resultados.

In [112]:
# Iterar sobre las filas y columnas del DataFrame 'res'
for index, row in res.iterrows():
    for col in ['LR', 'SVC', 'XGB']:  # Iterar sobre las columnas LR, SVC y XGB
        col_index = ['LR', 'SVC', 'XGB'].index(col)
        # Obtener el valor de 'key' basado en el valor en la columna 'col'
        key = row[col]
        # Obtener el valor asociado de 'pred_probs' para la fila 'index' y columna 'key'
        value = pred_probs[key][index]
        value = value[col_index]
        # Actualizar la celda en el DataFrame con el valor 'L: value'
        res.at[index, col] = f"{key}: {value}"

# Mostrar los resultados actualizados
res

Home           Away                     LR  \
0  Bayern Munich    Real Madrid  A: 0.9992342079861941   
1       Dortmund      Paris S-G  A: 0.9979236052920363   
2      Paris S-G       Dortmund  A: 0.7224568609987156   
3    Real Madrid  Bayern Munich  A: 0.7742763397897036   

                      SVC                      XGB  
0   H: 0.4446658105986785  D: 0.006322417075439371  
1  H: 0.44466523310524053    A: 0.5220480870374677  
2   H: 0.4446683567050252  D: 0.011538765465155301  
3  H: 0.44466760629762464  D: 0.011759547461316027

Vemos que para nuestro modelo de Regresión Logística son claros ganadores el Real Madrid y el PSG. 

Por otro lado, en SVC las probabilidades están más igualdas pero el resultado es el mismo: pasan a la final el Real Madrid y el PSG.

Por último, XGBoost nos muestra claramente que el PSG pasa a la final, pero no está tan claro el resultado de los partidos del Bayern Munich y el Real Madrid. Analicemos más a fondo los resultados de estos dos partidos en nuestro modelo XGBoost.

In [113]:
print('Bayern Munich vs Real Madrid:')
print('H: ', pred_probs['H'][0][-1])
print('D: ', pred_probs['D'][0][-1])
print('A: ', pred_probs['A'][0][-1])

print('-' * 50)

print('Real Madrid vs Bayern Munich:')
print('H: ', pred_probs['H'][3][-1])
print('D: ', pred_probs['D'][3][-1])
print('A: ', pred_probs['A'][3][-1])

Bayern Munich vs Real Madrid:
H:  0.6657942770815379
D:  0.006322417075439371
A:  0.3278833058430229
--------------------------------------------------
Real Madrid vs Bayern Munich:
H:  0.7631860256349595
D:  0.011759547461316027
A:  0.22505442690372432


Podemos observar que en el primer partido gana el Bayern con un 66% y en el segundo partido gana el Real Madrid con el 76%. 


Por lo tanto, concluimos que todos nuestros modelo de clasificación nos dan el mismo resultado para la final: Real Madrid vs PSG. Veamos quién gana esta final.